### PR TITLE
support NFS sharing with the hyperkit driver

### PR DESCRIPTION
**Pull request by mo-hit**

### CODE REVIEW

**Comment by dlorenc** [comment]
really cool!

@minikube-bot test this please

**Comment by mo-hit** [comment]
@dlorenc any insight on those tests?

**Comment by mo-hit** [comment]
had an extra log output I was using for debugging, removed that 😅

**Comment by r2d4** [comment]
@minikube-bot ok to test

**Comment by mo-hit** [comment]
@r2d4 seems like the linux tests are stalled? can't find my build artifact?

**Comment by dlorenc** [comment]
@minikube-bot test this please

File: vendor/vendor.json [hunk 1]
```diff
@@ -0,0 +1,25 @@
+{
```
Comment (dlorenc) [inline vendor/vendor.json]: What did you use to generate this file?

File: pkg/minikube/cluster/types.go [hunk 2]
```diff
@@ -23,21 +23,23 @@ import (
...
+	DockerEnv             []string // Each entry is formatted as KEY=VALUE.
+	InsecureRegistry      []string
+	RegistryMirror        []string
+	HostOnlyCIDR          string // Only used by the virtualbox driver
+	HypervVirtualSwitch   string
+	KvmNetwork            string             // Only used by the KVM driver
+	Downloader            util.ISODownloader `json:"-"`
+	DockerOpt             []string           // Each entry is formatted as KEY=VALUE.
+	DisableDriverMounts   bool               // Only used by virtualbox and xhyve
+	HyperkitNFSShare      []string
```
Comment (dlorenc) [inline pkg/minikube/cluster/types.go]: This is really cool, but most of the code doesn't actually seem hyperkit specific (just a big OSX specific). What do you think about dropping Hyperkit from the flag so we can more easily support the rest of the drivers and platforms?

**Comment by mo-hit** [comment]
@dlorenc sounds good, I'll remove the hyperkit prefix :)
I'm not 100% sure about the file to be hinest, still kinda new to vendoring / go. I'll investigate

**Comment by mo-hit** [comment]
@dlorenc I added those changes, fixed my vendoring issues. Those macOS/linux tests still seem to not be able to find the correct build artifact

**Comment by r2d4** [comment]
@mo-hit don't worry about the label, thats just automatically added by the k8s-bot.  The macOS tests should be working fine, lets see how they do this round.

File: pkg/drivers/hyperkit/driver.go [hunk 3]
```diff
@@ -207,6 +213,18 @@ func (d *Driver) Start() error {
 	if err := commonutil.RetryAfter(30, getIP, 2*time.Second); err != nil {
 		return fmt.Errorf("IP address never found in dhcp leases file %v", err)
 	}
+
+	if len(d.NFSShares) > 0 {
+		log.Info("Setting up NFS mounts")
+		// takes some time here for ssh / nfsd to work properly
+		time.Sleep(time.Second * 30)
```
Comment (dlorenc) [inline pkg/drivers/hyperkit/driver.go]: Could we poll/retry on an error instead of waiting 30s every time?

File: pkg/drivers/hyperkit/driver.go [hunk 4]
```diff
@@ -207,6 +213,18 @@ func (d *Driver) Start() error {
...
+		err = d.setupNFSShare()
```
Comment (dlorenc) [inline pkg/drivers/hyperkit/driver.go]: Nit: inline the err:= with the if below.

File: pkg/drivers/hyperkit/driver.go [hunk 5]
```diff
@@ -232,6 +250,56 @@ func (d *Driver) extractKernel(isoPath string) error {
...
+	if err != nil {
+		return err
+	}
+
+	hostIP, err := GetNetAddr()
+	if err != nil {
+		return err
+	}
+
+	mountCommands := fmt.Sprintf("#/bin/bash\\n")
```
Comment (dlorenc) [inline pkg/drivers/hyperkit/driver.go]: We probably want a "set -e" in there, otherwise it'll continue in the case of an error.

**Comment by mo-hit** [comment]
@dlorenc I've put my changes at https://github.com/mo-hit/minikube/tree/nfs-mount when you get some time :)

**Comment by r2d4** [comment]
fixes #2310

**Comment by dlorenc** [comment]
Thanks for adding me, I'll take a look tonight.

**Comment by huguesalary** [comment]
Thanks @mo-hit for working on this!

There's one small issue with the code that I'm trying to fix on my end:

The NFS exports need to be cleaned up (i.e. removed) from `/etc/exports` after `minikube stop` and `minikube delete`, otherwise the NFS configuration will refer to an invalid IP address (the IP of the previous `minikube start`).

**Comment by dlorenc** [comment]
Closing as this has been merged in another PR.

@huguesalary would you mind sending that PR directly here?

**Comment by huguesalary** [comment]
Sure thing, here's the PR: https://github.com/kubernetes/minikube/pull/2385

**Comment by mo-hit** [comment]
oooo right, It looks like my hyperkit-mount-fix branch was merged and not the nfs-mount one, I added some cleanup in the start here https://github.com/mo-hit/minikube/commit/244bf2d978e1af261ce8b40e260195d9d58b810f but yours looks nicer :)

**Pull request closed without merging**